### PR TITLE
Add @SpanAttribute annotation for adding attributes to spans created through instrumentation

### PR DESCRIPTION
**Pull request by HaloFour**
Adds a new annotated that when combined with the `@WithSpan` annotation will add the value of the annotated parameter as an attribute to the newly-created `Span`:

```java
@WithSpan("method")
public void someMethod(@SpanAttribute("name") String someName) {
    // do stuff
}
```

For the above example the OTel auto-instrumentation will create a new `Span` with the name "method" and will take the value of the parameter `someName` and add it as a String attribute with the name "name".

For the attribute names there is the caveat that, by default, the Java compiler does not include the parameter name metadata in the compiled classes.  You can opt-into this by passing the `-parameters` compiler argument.  Without this the instrumentation cannot derive a useful name for the parameter.  This could be an argument to make the `value` element of the annotation not have a default value.

See: https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/3188

### CODE REVIEW

File: extensions/annotations/src/main/java/io/opentelemetry/extension/annotations/WithSpanAttribute.java
Comment (Oberon00) [inline extensions/annotations/src/main/java/io/opentelemetry/extension/annotations/WithSpanAttribute.java:29]: Some more bikeshedding suggestions: `CaptureAsSpanAttribute`, `AsSpanAttribute`, `AddToSpan`

File: extensions/annotations/src/main/java/io/opentelemetry/extension/annotations/WithSpanAttribute.java
Comment (Oberon00) [inline extensions/annotations/src/main/java/io/opentelemetry/extension/annotations/WithSpanAttribute.java:33]: Since this can be used on multiple parameters, wouldn't that be more `classname.method.argname`? Or should this be a required parameter?

File: extensions/annotations/src/main/java/io/opentelemetry/extension/annotations/WithSpanAttribute.java
Comment (Oberon00) [inline extensions/annotations/src/main/java/io/opentelemetry/extension/annotations/WithSpanAttribute.java:34]:
```suggestion
   * E.g. {@code className.method}
```

File: extensions/annotations/src/main/java/io/opentelemetry/extension/annotations/WithSpanAttribute.java
Comment (HaloFour) [reply extensions/annotations/src/main/java/io/opentelemetry/extension/annotations/WithSpanAttribute.java:33]: oops, that is total copypasta on my part.  I would think that the parameter name itself would be sufficient, although note the caveat mentioned in above that this would require that the code be compiled with `-parameters` which should definitely be mentioned in this javadoc comment.

**Comment by jkwatson** [comment]
Why is this a better approach than just calling `Span.current().setAttribute(...)` ? The advantage of `@WithSpan` is that it hides a bunch of boilerplate try-with-resources/finally code. This, in my opinion, just adds some hard-to-read annotation gunk, which clutters up method signatures, rather than providing any real value.

**Comment by jkwatson** [comment]
> Why is this a better approach than just calling `Span.current().setAttribute(...)` ? The advantage of `@WithSpan` is that it hides a bunch of boilerplate try-with-resources/finally code. This, in my opinion, just adds some hard-to-read annotation gunk, which clutters up method signatures, rather than providing any real value.

Just adding to this, if a user chose to use these annotations, they would be requiring that the implementation do the non-optimal thing of creating a new instance of the AttributeKey for every call, rather than allocating them up front (or, even better, using the semantic convention attributes).

**Comment by HaloFour** [comment]
@jkwatson 

> Why is this a better approach than just calling `Span.current().setAttribute(...)` ?

It's a minor point but the instrumentation can choose to not create a span and that would have to be something that the code within the method then attempts to detect.  But that is probably a fairly rare circumstance.  I think the only time it comes up now is if you try to create a `SERVER` or `CLIENT` span when already in the context of another span of the same kind.

> Just adding to this, if a user chose to use these annotations, they would be requiring that the implementation do the non-optimal thing of creating a new instance of the AttributeKey for every call, rather than allocating them up front (or, even better, using the semantic convention attributes).

This is a good point.  The implementation I have in my own aspect memoizes the "bindings" to the method parameters so the `AttributeKey<?>` instances are created only once per method.  I've not gotten that far with the draft PR to OTel Instrumentation.  But you couldn't use the shared semconv attribute key instances.  I don't know that this is a terribly big deal as the cases where I've used this kind of functionality has been for more custom attributes that I wouldn't expect to live in the semconv module.

As for the general usefulness of this feature, I leave that to the group to decide.  There seemed to be some interest when I posted a [discussion](https://github.com/open-telemetry/opentelemetry-java-instrumentation/discussions/2037) that mentioned it.  But it is admittedly a pretty minor feature.

**Comment by Oberon00** [comment]
Another advantage of annotations is that you only need a `compileOnly` dependency on them. So if you don't need the otel-api otherwise, you have one dependency fewer.

File: extensions/annotations/src/main/java/io/opentelemetry/extension/annotations/WithSpanAttribute.java
Comment (jkwatson) [reply extensions/annotations/src/main/java/io/opentelemetry/extension/annotations/WithSpanAttribute.java:29]: If we go with this addition, I'd just have it be `@SpanAttribute`. It's going to be super noisy in the source code already, so we should strive to make it as concise as is reasonable.

Comment (anuraaga) [reply extensions/annotations/src/main/java/io/opentelemetry/extension/annotations/WithSpanAttribute.java:29]: +1 to `SpanAttribute`

File: extensions/annotations/src/main/java/io/opentelemetry/extension/annotations/WithSpanAttribute.java
Comment (anuraaga) [reply extensions/annotations/src/main/java/io/opentelemetry/extension/annotations/WithSpanAttribute.java:33]: We should phrase our javadoc for the end user, not instrumentation authors, in general I think (if notes for instrumentation are appropriate, I'd rather have them in a .md file than javadoc). So this would be something like

```
If not specified and the code is compiled using the `{@code -parameters}` argument to `javac`, the parameter name will be used instead. If the parameter name is not available, e.g., because the code was not compiled with that flag, the attribute will be ignored.
```

**Review by anuraaga — COMMENTED** [review]
Now I've posted on the linked discussion, and the Slack, and here - hope things didn't fragment too much :)

As I posted in the discussion, I find supporting this sort of annotation idiomatic vs frameworks like spring. I think the performance issue is a non-issue since as posted in slack, there are cardinality guarantees that mean we should just cache the attribute keys (with bytecode manipulation, technically it doesn't even need to be a Map, the keys could be added to a static initializer and referenced directly. We have no tooling for such coolness though and would probably just use a Map at first).

> Another advantage of annotations is that you only need a compileOnly dependency on them. So if you don't need the otel-api otherwise, you have one dependency fewer.

In user apps this may commonly be the case, but let's not forget our own troubles with `compileOnly` and `GuardedBy` :) In principle, we should assume anything that's not `SOURCE` retention does require a proper dependency.

**Comment by jkwatson** [comment]
I don't love this, because I feel that annotation-oriented-programming is not a great approach to writing readable code, but if the other maintainers are on board, I won't stand in the way.

File: extensions/annotations/src/main/java/io/opentelemetry/extension/annotations/WithSpanAttribute.java
Comment (mateuszrzeszutek) [inline extensions/annotations/src/main/java/io/opentelemetry/extension/annotations/WithSpanAttribute.java:25]: We shouldn't use the old java instrumentation repo URL:

```suggestion
 * @see <a href="https://github.com/open-telemetry/opentelemetry-java-instrumentation">OpenTelemetry
 *     OpenTelemetry Instrumentation for Java</a>
```

**Comment by trask** [comment]
I think this is a nice option for some uses. I think it's worth putting the alternative `Span.current().setAttribute(...)` into the `@SpanAttribute` javadoc, since it's not uncommon to want to pull out a nested field from a parameter. Also that should prevent people from getting the idea that this is the only want to capture span attributes, and then adding method parameters just so that they can be captured with `@SpanAttribute` 😄.

I'm not sure about making the attribute name optional, it seems too easy to break something, e.g. by compiling the class differently, or by refactoring and renaming the parameter. Are there examples of other annotations that do this (e.g. JAX-RS, etc)?

**Comment by HaloFour** [comment]
@trask 

> I'm not sure about making the attribute name optional, it seems too easy to break something, e.g. by compiling the class differently, or by refactoring and renaming the parameter. Are there examples of other annotations that do this (e.g. JAX-RS, etc)?

The Spring parameter mapping annotations (`@PathVariable`, `@RequestParam`) have optional names and will infer the name from the parameter.

AFAICT the JAXRS annotations all require the name.

File: extensions/annotations/src/main/java/io/opentelemetry/extension/annotations/WithSpanAttribute.java
Comment (HaloFour) [reply extensions/annotations/src/main/java/io/opentelemetry/extension/annotations/WithSpanAttribute.java:25]: Oops, also a copypasta.  I'll update the link in the javadoc for `@WithSpan` while I'm at it.

**Comment by kubawach** [comment]
> @trask
> 
> > I'm not sure about making the attribute name optional, it seems too easy to break something, e.g. by compiling the class differently, or by refactoring and renaming the parameter. Are there examples of other annotations that do this (e.g. JAX-RS, etc)?
> 
> The Spring parameter mapping annotations (`@PathVariable`, `@RequestParam`) have optional names and will infer the name from the parameter.
> 
> AFAICT the JAXRS annotations all require the name.

I think that optional names strike a good compromise between necessity to have a constant attribute name (cause eg automation relies on it) and moving fast with the implementation on the customer side (names of attributes same as parameter - less friction).

**Comment by kubawach** [comment]
From my point of view this is an interesting feature, worth implementing. Coming from enterprise software I often saw that annotation driven approach to implementing aspects of an application not directly related to the core business functionality (such as logging, telemetry, security...) is very welcomed. Therefore this kind of feature can increase / speed-up adoption of OTEL :)

File: extensions/annotations/src/main/java/io/opentelemetry/extension/annotations/SpanAttribute.java
Comment (kubawach) [inline extensions/annotations/src/main/java/io/opentelemetry/extension/annotations/SpanAttribute.java:19]: This seems a copy-paste :) Probably should read: 
```
* <p>Application developers can use this annotation to signal OpenTelemetry auto-instrumentation
*  that a new span attribute should be added to a span created when the parent method is executed.
```

**Review by kubawach — CHANGES_REQUESTED** [review]

File: extensions/annotations/src/main/java/io/opentelemetry/extension/annotations/SpanAttribute.java
Comment (HaloFour) [reply extensions/annotations/src/main/java/io/opentelemetry/extension/annotations/SpanAttribute.java:19]: Good catch, indeed a copypasta fail!

**Review by anuraaga — APPROVED** [review]
We discussed at SIG today and decided on optional params. The main motivation is that we found Spring has many mechanisms for resolving names, including a free one for kotlin users. We're hoping you'll help us implement these :)

**Comment by HaloFour** [comment]
Sounds good.  I'm going to try to put some effort into the instrumentation side of things this weekend.  Based on the conversations here I want to ensure that the implementation keeps the cost down as much as possible, e.g. ensuring that the `AttributeKey<T>` is allocated exactly one.

I think there's a conversation around how flexible the instrumentation should attempt to be when coercing a parameter type to an attribute where it would require allocations or copying.  For example, to me it feels reasonable to support parameters of type `long[]`, but there are no free ways to get that into a `List<Long>` compatible with the OTel API.

**Review by trask — APPROVED** [review]

File: extensions/annotations/src/main/java/io/opentelemetry/extension/annotations/SpanAttribute.java
Comment (jkwatson) [inline extensions/annotations/src/main/java/io/opentelemetry/extension/annotations/SpanAttribute.java:31]: We need a `@since 1.2.0` on this class javadoc as well, if we're going to add it.

**Comment by jkwatson** [comment]
@bogdandrutu since this is effectively an API addition, if only in the annotation package, your opinion here matters as well.

File: extensions/annotations/src/main/java/io/opentelemetry/extension/annotations/SpanAttribute.java
Comment (anuraaga) [reply extensions/annotations/src/main/java/io/opentelemetry/extension/annotations/SpanAttribute.java:31]: @HaloFour Once you add this, we'll merge :)

Comment (HaloFour) [reply extensions/annotations/src/main/java/io/opentelemetry/extension/annotations/SpanAttribute.java:32]: Sorry about that, been a bit swamped recently.  Speaking of, with this merged is there an expected turnaround time for instrumentation to match this annotation?  I've been working on some changes to my WIP PR to try to address some of the concerns brought up here but it has some more to go and I've not had a lot of opportunity to look at it this week.

Comment (trask) [reply extensions/annotations/src/main/java/io/opentelemetry/extension/annotations/SpanAttribute.java:32]: > Speaking of, with this merged is there an expected turnaround time for instrumentation to match this annotation?

that's a good thought. I hadn't thought about it, but it would be ideal for this annotation to land in the same release cycle as the support for it in the java agent to avoid user confusion (this repo usually releases first week of the month, and instrumentation repo usually releases the week after that).

makes sense to hold off on merging this PR for now then?

Comment (jkwatson) [reply extensions/annotations/src/main/java/io/opentelemetry/extension/annotations/SpanAttribute.java:32]: I'm happy to hold off until after the release. Of course, I'd be happy to wait even longer. ;)

Comment (HaloFour) [reply extensions/annotations/src/main/java/io/opentelemetry/extension/annotations/SpanAttribute.java:32]: @trask 

> makes sense to hold off on merging this PR for now then?

That works for me.

@jkwatson 

>  Of course, I'd be happy to wait even longer. ;)

You should see the projects I've been working on lately.  Spring WebFlux + Swagger/OpenAPI.  I think we have twice as many annotations as we have code.  😁

Comment (jkwatson) [reply extensions/annotations/src/main/java/io/opentelemetry/extension/annotations/SpanAttribute.java:32]: Swagger is the worst. It's the example I always cite about annotation abuse.

**Comment by jkwatson** [comment]
@HaloFour I think this needs a rebase to pick up the changes to our CI settings to do a build on 20.04.

**Comment by HaloFour** [comment]
Sorry about the holdup, life got in the way.  I've rebased this PR and I've also pushed up a draft PR of the implementation with the ByteBuddy instrumentation attempting to simplify the approach and take into consideration some of the concerns brought up here and on Slack:  https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/3188

**Comment by jkwatson** [comment]
Since we're planning on a release next week, should we hold off until just after that release, so the instrumentation project has a chance to integrate?

**Comment by HaloFour** [comment]
@jkwatson 

> Since we're planning on a release next week, should we hold off until just after that release, so the instrumentation project has a chance to integrate?

Agreed.  I'd want to ensure that the instrumentation PR gets plenty of careful consideration and while I've been making progress I've only recently marked it ready for review.  I'd be happy to wait until that PR has been approved (or as close as it can be without this annotation being ready).

**Review by jkwatson — APPROVED** [review]
sigh

**Review by anuraaga — APPROVED** [review]

**Comment by anuraaga** [comment]
@HaloFour Don't mind @jkwatson :P Thanks!

**Pull request merged**
